Feat(web-react): Introduce Uncontrollable Toast

pavelklibani opened the pull request:
## Introduce `<UncontrolledToast />`

Storybook:
https://deploy-preview-1344--spirit-design-system-storybook.netlify.app/?path=/story/components-toast--uncontrolled-toast-playground

literat commented on packages/web-react/src/components/Toast/UncontrolledToastContext.tsx:
In general, the `Uncontrolled` should be only the component. Other functionality, like a hook or context, should not use this prefix in a file name. Image that you want to use the Toast functionality and assemble this by yourself, so will be using `ToastContext` as well as calling the `useToast` hook. The `UncontrolledToast` means that this all is done in this component under the hood but hooks and context can be used also separately.

I suggest renaming those files without `uncontrolled`.

pavelklibani replied:
you are right, context component is renamed.

literat commented on packages/web-react/src/types/toast.ts:
```diff
@@ -6,6 +6,8 @@ import {
   StyleProps,
 } from './shared';
 
+export type ToastColorType = 'inverted' | EmotionColorsDictionaryType;
```
👍

pavelklibani commented on packages/web-react/src/components/Toast/UncontrolledToast.tsx:
```diff
@@ -0,0 +1,57 @@
+import { useUncontrolledToast } from './useUncontrolledToast';
+import { ChildrenProps, StyleProps } from '../../types';
+
+interface UncontrolledToastProps extends ChildrenProps, StyleProps {
```
moved to types

crishpeen commented on packages/web-react/src/components/Toast/stories/UncontrolledToast.stories.tsx:
```diff
@@ -0,0 +1,107 @@
+    },
+    color: {
+      control: 'select',
+      options: ['inverted', 'danger', 'warning', 'success', 'info'],
```
```suggestion
      options: ['inverted', ...Object.values(EmotionColors)],
```
can we do this? Also in the toastbar story.

pavelklibani replied:
we can do this! thanks

crishpeen commented on packages/web-react/src/components/Toast/stories/UncontrolledToast.stories.tsx:
```diff
@@ -0,0 +1,107 @@
+    },
+    alignmentX: {
+      control: 'select',
+      options: ['left', 'center', 'right'],
```
```suggestion
      options: [...Object.values(AlignmentX)],
```

pavelklibani replied:
changed

crishpeen commented on packages/web-react/src/components/Toast/stories/UncontrolledToast.stories.tsx:
```diff
@@ -0,0 +1,107 @@
+    },
+  },
+  args: {
+    children: 'Hello, World! This is a toast message with an action.',
```
does it actually have an action?

pavelklibani replied:
there is no action - removed

crishpeen commented on packages/web-react/src/components/Toast/README.md:
```diff
@@ -263,13 +264,108 @@ const [isOpen, setIsOpen] = React.useState(false)
+
+| Name        | Type                                                         | Default    | Description                                              |
+| ----------- | ------------------------------------------------------------ | ---------- | -------------------------------------------------------- |
+| `autoClose` | `bool`                                                       | `false`    | Turn on/off auto close of UncontrolledToast after 5000ms |
```
whoa, we don't have this feature in the web package, right? @adamkudrna

literat replied:
We agreed with PM, that there is no demand for this feature so it should be removed.

pavelklibani replied:
After discussion with @literat and Frantisek, we decided to drop this feature.

crishpeen commented on packages/web-react/src/components/Toast/README.md:
```diff
@@ -263,13 +264,108 @@ const [isOpen, setIsOpen] = React.useState(false)
+    Show Toast
+  </Button>
+
+  <UncontrolledToast alignmentX="right" alignmentY="top" closeLabel="Close toast" hasIcon isDismissible />
```
message?

pavelklibani replied:
message added

pavelklibani replied:
Reverting back, I just realized there should not be any message because the message is dynamic and you are setting it with the call function. So the message is added to this component from context.

crishpeen commented on packages/web-react/src/components/Toast/README.md:
```diff
@@ -263,13 +264,108 @@ const [isOpen, setIsOpen] = React.useState(false)
+
+This hook returns:
+
+| Name        | Type                                                         | Default    | Description                                              |
```
no required column?

pavelklibani replied:
No, this is the table for `useUncontrolledToast` hook. 
So you can understand how to manage this component and what data you can retrieve from it.

Do we have a better way to pass that info to the developer?

crishpeen commented on packages/web-react/src/components/Toast/README.md:
```diff
@@ -263,13 +264,108 @@ const [isOpen, setIsOpen] = React.useState(false)
+              │                 ┌─⫸ ToastBar ID (required)
+              │                 │
+showToast('Toast message', 'toast-id', {
+  autoClose: true         // Automatically close toast after 5000ms, default: false
```
you are probably missing commas in this object.

pavelklibani replied:
commas added

crishpeen commented on packages/web-react/src/components/Toast/README.md:
```diff
@@ -263,13 +264,108 @@ const [isOpen, setIsOpen] = React.useState(false)
+| `hasIcon`       | `bool`                                                      | `false` \* | ✕        | If true, an icon is shown along the message |
+| `isDismissible` | `bool`                                                      | `false`    | ✕        | If true, ToastBar can be dismissed by user  |
+
+(\*) For each emotion color, a default icon is defined.
```
am I not able to change this icon?

pavelklibani replied:
you can change it with call function `showToast` from `useUncontrolledToast` same as `color` and `autoClose`

<img width="854" alt="image" src="https://github.com/lmc-eu/spirit-design-system/assets/26870554/60045230-8760-4bbb-88d1-ba8bfac56eaf">

literat commented on packages/web-react/src/components/Toast/useToastIcon.ts:
`.jsx/.tsx` means that the file includes `JSX` syntax.

literat commented on packages/web-react/src/components/Toast/useToast.ts:
```diff
@@ -0,0 +1,12 @@
+  const currentToastContext = useContext(ToastContext);
+
+  if (!currentToastContext) {
+    throw new Error('useToast has to be used within <useToast>');
```
Did you mean `ToastProvider`?

```suggestion
    throw new Error('useToast has to be used within <ToastProvider>');
```

pavelklibani replied:
yes, my mistake, thank you!

literat commented on packages/web-react/src/components/Toast/index.ts:
Shouldn't we export also `useToast` hook?

pavelklibani replied:
added to exports

pavelklibani commented on packages/web-react/src/components/Toast/useUncontrolledToast.ts:
renamed to `useToast`

pavelklibani commented:
Since we dropped the `autoClose` feature, the only way how to close `UncontrolledToast` is close button inside the UncontrolledToast (`isDismissible`). So i changed this prop default to true.

Another way is to take `close` function from `useToast` hook and call it.

My question is - now there are functions `showToast` and `close`... I would like to rename them.
I suggest `open`/`close`, `showToast`/`hideToast`, `openToast`/`closeToast`

What do you think @literat ?

literat reviewed (approved):
Good Job. LGTM 👍
